example_cpp_smart_card_client_app/src/pp_module.cc

```diff
@@ -54,6 +54,56 @@ namespace smart_card_client {
 
 namespace {
 
+// Collects all currently available certificates.
+// Returns whether the operation finished successfully. In case of success,
+// the resulting certificates information should be returned through the
+// |certificates| output argument.
+bool GetCertificates(std::vector<ccp::ClientCertificateInfo>* certificates) {
+  //
+  // CHANGE HERE:
+  // Place your custom code here:
+  //
+
+  // Note: the bytes "1, 2, 3" and the signature algorithms below are just
+  // an example. In the real application, replace them with the bytes of the
+  // DER encoding of a X.509 certificate and the supported algorithms.
+  ccp::ClientCertificateInfo certificate_info_1;
+  certificate_info_1.certificate.assign({1, 2, 3});
+  certificate_info_1.supported_algorithms.push_back(
+      ccp::Algorithm::kRsassaPkcs1v15Sha1);
+  ccp::ClientCertificateInfo certificate_info_2;
+  certificate_info_2.supported_algorithms.push_back(
+      ccp::Algorithm::kRsassaPkcs1v15Sha512);
+  certificates->push_back(certificate_info_1);
+  certificates->push_back(certificate_info_2);
+  return true;
+}
+
+// Reports all currently available certificates to Chrome via the
+// chrome.certificateProvider.setCertificates API. Returns false if this failed.
+bool ReportAvailableCertificates(
+    std::weak_ptr<ccp::ApiBridge> chrome_certificate_provider_api_bridge) {
+  const std::shared_ptr<ccp::ApiBridge>
+      locked_chrome_certificate_provider_api_bridge =
+          chrome_certificate_provider_api_bridge.lock();
+  if (!locked_chrome_certificate_provider_api_bridge) {
+    GOOGLE_SMART_CARD_LOG_INFO << "Cannot provide certificates: The " <<
+        "shutdown process has started";
+    return false;
+  }
+
+  std::vector<ccp::ClientCertificateInfo> certificates;
+  if (GetCertificates(&certificates)) {
```

Comment:
So when this function fails, we won't notify Chrome with an empty list? I guess this might be useful in case of some transient error. But that shouldn't be used in case the failure is simply "there's no smart card available", since instead of immediately notifying Chrome about the empty list we'll wait forever.

I'm concerned that the extension author might confuse these two situations and cause weird side effects. So I'd lead towards dropping the bool result, and only return the vector from the function. WDYT?

Reply:
Done